fvm/src/machine/mod.rs

```diff
@@ -97,6 +98,10 @@ pub trait Machine: 'static {
 
     /// Creates a new limiter to track the resources of a message execution.
     fn new_limiter(&self) -> Self::Limiter;
+
+    /// Commits the events to the machine by building the events AMT, and making sure that events
+    /// are written to the store.
+    fn commit_events(&self, events: &[StampedEvent]) -> Result<Option<Cid>>;
```

Comment:
In the future it will, though. We are planning to fold events into a block-level index structure at some point. My understanding is that the logic to populate that data structure would go here, and be returned when finalizing the Machine.

Reply:
Hm. I _assume_ that logic would go in the client. But I could see it going here as well.

Reply:
(basically, the FVM usually deals with the state-tree only)

Reply:
Assuming that we _don't_ return events to the client in ApplyRet (which we're doing now, but because to support transient call cases, and not sure if it needs to be there, might remove), doing it there would involve the extra work of re-fetching the events from the store _just_ to populate the structure.

FWIW, I think forming the receipts HAMT should _also_ be done in the Machine, with us returning the receipt_root to the client. IMO we whould trend towards consolidating all consensus-relevant execution logic inside the FVM. Right now there is still a gray area, but I'm strongly advocating for that future event indexing logic to live here.

Reply:
I'm strongly against it. In general, the FVM should _just_ be the VM. It manages the state-tree, applies messages, and returns receipts. It doesn't care about the receipt HAMT, tipsets, the message lists, etc.

1. Different chains (e.g., in IPC) may want to use the FVM but may also have different chain structures and indexing requirements. E.g., I expect some chains will use cryptographic accumulators, vector commitments, etc.
2. Filecoin clients will likely want to index events in some form of database _anyways_, so it's not like they won't have to process them.